### PR TITLE
ghdl: add libbacktrace for llvm backend

### PR DESCRIPTION
**Pull request by Aloz1**
#### Testing the changes
- I tested the changes in this PR: **briefly**

#### Local build testing
- I built this PR locally for my native architecture, (ARCH-LIBC)

### CODE REVIEW

**Comment by Aloz1** [comment]
Would like some feedback on how I've implemented this.

At present, I've added libbacktrace.a to libgcc-devel, and ghdl uses that for the llvm backend (it just copies it). It irks me a bit that it copies libbacktrace, and I'm also not happy with the way I've got the template detecting where to find libbacktrace.a

A few possible solutions:
1) Fix detection of /usr/lib/gcc/<target>/<minver> for cross builds, maybe update the libbacktrace copy to a symlink while we're at it.
2) Use [ianlancetaylor/libbacktrace](/ianlancetaylor/libbacktrace) instead (they unfortunately don't seem to do any versioning at all).
3) Download gcc source and build libbacktrace from that (gcc source will eventually be needed anyway for building ghdl-gcc)

Any thoughts?

File: srcpkgs/gcc/template [hunk 1]
```diff
@@ -353,6 +353,10 @@ do_install() {
 	ln -sfr ${DESTDIR}/usr/include/c++/${_minorver} \
 		${DESTDIR}/usr/include/c++/${_patchver}
 
+	# Install libbacktrace. (it's not installed by makefiles)
+	vinstall libbacktrace/.libs/libbacktrace.a \
+		644 usr/lib/gcc/${_triplet}/${_minorver}
+
```
Comment (ericonr) [inline srcpkgs/gcc/template:359]: This should be its own commit.

I am not enthused about installing something that comes from a `.libs` directory into the package. Are there even headers for this?

File: srcpkgs/ghdl/template [hunk 2]
```diff
@@ -63,9 +63,15 @@ do_configure() {
 		cd ..
 	fi
 	if [ "$build_option_llvm" ]; then
+		# FIXME: This is a bit hacky. This should be replaced by the path within the gcc
+		# source tree for libbacktrace when building the gcc backend has been added.
+		gcc_dir=$(gcc --print-search-dirs | awk '{ if ($1 == "install:") print $2 }')
```
Comment (ericonr) [inline srcpkgs/ghdl/template:68]: Doing this like we do in `python3-pyside` is better. You can just build the library path yourself.

`		local gcc_ver="$(${XBPS_CROSS_TRIPLET}-gcc -dumpversion)"`

File: srcpkgs/ghdl/template [hunk 3]
```diff
@@ -63,9 +63,15 @@ do_configure() {
...
+
 		mkdir -p build_llvm
 		cd build_llvm
-		../configure --libdir=lib/ghdl/llvm --with-llvm-config \
+		../configure --libdir=lib/ghdl/llvm \
+			--with-llvm-config \
+			--with-backtrace-lib="${XBPS_MASTERDIR}${gcc_dir}/libbacktrace.a" \
```
Comment (ericonr) [inline srcpkgs/ghdl/template:74]: masterdir doesn't fit here.

File: srcpkgs/gcc/template (hunk 1 above)
Comment (Aloz1) [reply srcpkgs/gcc/template:359]: Well, not even sure if this is the route I want to take. I'd much prefer upstream libbacktrace, but upstream doesn't do versioning (no tags, no releases, just a master branch on a git repo).

Yep, I know, but the gcc source doesn't install libbacktrace at all. This is inspired by how Debian handle libbacktrace. 

Yes, headers are available...sorry, my bad. Will include them.

Any thoughts regarding my request for feedback?

File: srcpkgs/ghdl/template (hunk 2 above)
Comment (Aloz1) [reply srcpkgs/ghdl/template:68]: Ah neat. Thanks for that.

File: srcpkgs/ghdl/template (hunk 3 above)
Comment (Aloz1) [reply srcpkgs/ghdl/template]: Ah yep, fixed.

**Comment by ericonr** [comment]
What does ghdl use libbacktrace for, specifically?

> Use ianlancetaylor/libbacktrace instead (they unfortunately don't seem to do any versioning at all).

I think using GCC's, since it's already there, is "reasonable". But it should be its own subpackage, the ugliness should be opt-in.

> Download gcc source and build libbacktrace from that (gcc source will eventually be needed anyway for building ghdl-gcc)

What does it need GCC for?

**Comment by Aloz1** [comment]
> What does ghdl use libbacktrace for, specifically?

GHDL is a VHDL simulator (well, more like an ahead of time compiler nowadays). It links in libbacktrace to elaborated (compiled) VHDL designs. When an error or warning occurs, a backtrace is generated by the program. It actually only needs libbacktrace for the llvm backend (libbacktrace is already linked into libgcc).

> > Use ianlancetaylor/libbacktrace instead (they unfortunately don't seem to do any versioning at all).
> 
> I think using GCC's, since it's already there, is "reasonable". But it should be its own subpackage, the ugliness should be opt-in.

OK, cool. Makes total sense

> > Download gcc source and build libbacktrace from that (gcc source will eventually be needed anyway for building ghdl-gcc)
> 
> What does it need GCC for?

Because GHDL is essentially a compiler for VHDL, it generates machine code using one of 3 backends (mcode, llvm, or gcc).
In other words, GHDL is essentially a VHDL fronted for LLVM or GCC.

At the moment, only the LLVM and mcode variants of GHDL are being built. To build the GCC variant, it requires a source distribution of GCC (which, AFAIK, is the same as for any other custom language frontend for GCC).

So the rationale here is that, if we build ghdl-gcc, then llvm could use the libbacktrace that gets built as part of the gcc source distribution instead of packaging the one that gets built as part of the gcc package.

**Comment by ericonr** [comment]
> When an error or warning occurs, a backtrace is generated by the program. It actually only needs libbacktrace for the llvm backend (libbacktrace is already linked into libgcc).

What isn't clear to me is: does the current version not work at all, or is it simply missing the generated backtraces?

> At the moment, only the LLVM and mcode variants of GHDL are being built. To build the GCC variant, it requires a source distribution of GCC (which, AFAIK, is the same as for any other custom language frontend for GCC).

Unless it provides specific functionality, I'd rather avoid that extra backend :P 

> So the rationale here is that, if we build ghdl-gcc, then llvm could use the libbacktrace that gets built as part of the gcc source distribution instead of packaging the one that gets built as part of the gcc package.

That sounds messy but workable, I suppose. This whole thing seems kinda wild tbh ;-;

**Comment by Aloz1** [comment]
> What isn't clear to me is: does the current version not work at all, or is it simply missing the generated backtraces?

Oh, the current llvm backend works fine, it just doesn't produce backtraces, which makes debugging VHDL code quite difficult.

> Unless it provides specific functionality, I'd rather avoid that extra backend :P

As far as I can tell, the only feature that ghdl-gcc offers that other backends do not is code coverage (which is done via gcov). After mulling over this for a bit, its unimportant enough that I think dealing with the ghdl-gcc mess can be done later if absolutely necessary. I'm just using ghdl for some personal VHDL projects, so don't need code coverage.

> That sounds messy but workable, I suppose. This whole thing seems kinda wild tbh ;-;

Yeah, the whole method of using gcc as an intermediary for generating binaries is a lot more of a dogs breakfast than with llvm.

**Comment by ericonr** [comment]
> Oh, the current llvm backend works fine, it just doesn't produce backtraces, which makes debugging VHDL code quite difficult.

Understandable. How stable of an interface is `libbacktrace` inside of GCC? I'm weary about creating this dependency.

**Comment by Aloz1** [comment]
> Understandable. How stable of an interface is `libbacktrace` inside of GCC? I'm weary about creating this dependency.

Looking at git history for libbacktrace, it seems to be fairly stable. There are very few commits, and there seems to be only 1 (albeit fairly recent) breaking change in backtrace.h. It also seems that the gcc source libbacktrace and the upstream [ianlancetaylor/libbacktrace](/ianlancetaylor/libbacktrace) are kept relatively in-sync with each other.

**Comment by Aloz1** [comment]
> I am not enthused about installing something that comes from a .libs directory into the package.

I've now added a patch for configure.ac and Makefile.am (as well as re-genned configure and Makefile.in) so that the gcc build system is doing the heavy lifting. These changes mostly came from looking at a diff between gcc/libbacktrace and ianlancetaylor/libbacktrace.

> I think using GCC's, since it's already there, is "reasonable". But it should be its own subpackage, the ugliness should be opt-in.

libbacktrace from gcc now split into its own subpackage, and I've made it a separate commit as requested in one of your review comments.

**Comment by ericonr** [comment]
> I've now added a patch for configure.ac and Makefile.am (as well as re-genned configure and Makefile.in) so that the gcc build system is doing the heavy lifting. These changes mostly came from looking at a diff between gcc/libbacktrace and ianlancetaylor/libbacktrace.

The upstream build still puts it inside `.libs`, this doesn't change that. A 1k GCC patch is not the solution here, please revert to the version with the manual install commands.

**Comment by Aloz1** [comment]
Any further advice/comments thoughts on this one?
